---
title: rawrr - Functional test using raw files (autoQC01) from different LC-MS systems
author: 
  - Tobias Kockmann^[Functional Genomics Center Zurich (FGCZ), ETH Zurich / University of Zurich, Winterthurerstrasse 190, 8057 Zurich, Switzerland]
  - Christian Panse^[FGCZ and Swiss Institute of Bioinformatics, Quartier Sorge - Batiment Amphipole, 1015 Lausanne, Switzerland]
date: "`r Sys.time()`"
output:
  html_document:
    df_print: paged
    toc_float: true
    toc: true
    number_sections: true
    theme: united
bibliography: rawrr.bib
urlcolor: blue
abstract: |
  Thermo Fisher Scientific raw files were recorded across different MS devices using an internal quality control sample at the
  [FGCZ proteomics unit](https://fgcz.ch/omics_areas/prot.html)
  using the latest stable `R` and [rawrr](https://github.com/fgcz/rawrr)
  package versions.
  A Rmarkdown [script](https://github.com/fgcz/rawrr/blob/master/vignettes/functional_test.Rmd) process generates this report.
---

```{r sticker, out.width="25%", fig.align="center", fig.cap=" Lions love raw meat!", echo=FALSE}
knitr::include_graphics("images/hex-rmarkdown.png")
```

This R script processes raw files recorded across different Thermo Fisher Scientific Mass Spectrometers at the [FGCZ proteomics unit](https://fgcz.ch/omics_areas/prot.html) using the latest `R` development and [rawrr](https://github.com/fgcz/rawrr) package versions.

`<!--
TK: Asking myself if this couldn't be done using GitHub actions? It sounds sooooo much like continuous integration (CI).
-->

The main objectives are:

* Demonstrate code robustness with respect to different instrument models.
* Showcase dynamic application (daily changing input data).
* Demonstrate multicore computation by applying the `parallel` package.

The automatically updated output of this R script can be viewed at [https://fgcz-ms.uzh.ch/~cpanse/rawrr/test/functional_test.html]

Note: The analyzed sample (termed autoQC01) consists of the iRT peptide mix (Biognosys) in a tryptic BSA digest (NEB) and was separated applying a 20 min linear gradient on C18 RP material at a constant flow rate of 300 nl/min.


```{r}
library(rawrr)
library(parallel)

# only used for computing PSMs
library(protViz)
```


# Input data

A static snapshot of raw files was deposited as MassIVE dataset (MSV000086542)[https://massive.ucsd.edu/ProteoSAFe/dataset.jsp?accession=MSV000086542].
For local computation, please download the raw files using [ftp://massive.ucsd.edu/MSV000086542/].

Loads raw file listing.

```{r input}
input <- scan("input.txt", as.character())
```


```{r list}
knitr::kable(data.frame(Id=1:length(input), rawfile=input),
             caption="Input files available through MSV000086542.")
```

# Demonstrate `rawrr` reader functions on different Orbitrap models

Below we check the functionality of all reader functions of the `rawrr` package.

|Function Name     |Description                                             |Return value            |
|:-----------------|--------------------------------------------------------|:-----------------------|
|`readFileHeader`  |Reads meta information from a raw file header.           |`list`                  |
|`readIndex`       |Reads scan index from a raw file.                        |`data.frame`            |
|`readSpectrum`    |Reads spectral data from a raw file                      |`rawrrSpectrum(Set)`    |
|`readChromatogram`|Extracts chromatograms (TIC, BPC, or XIC) from a raw file|`rawrrChromatogram(Set)`|
Table: Overview of the _read_ functions of the `rawrr` package.

## Set number of concurrent running tasks

Set the number of `mc.cores` to `r detectCores()`.

```{r mccores}
options(mc.cores = detectCores())
options('mc.cores')
```

## `readFileHeader`

Read a list of meta data for each raw file.

```{r fileHeader}
H <- mclapply(FUN=rawrr::readFileHeader, input)
```

```{r defineInfo, echo=FALSE}
.getInstrumentInformation <- function(x){
  data.frame(
             model=x$`Instrument model`,
             serialNumber=x$`Serial number`,
             #method=x$`Instrument method`,
             softwareVersion=x$`Software version`,
             nMS=x$`Number of scans`,
             nMS2=x$`Number of ms2 scans`)
}
```

The table below lists some instrument information:

```{r knitr, results='asis'}
knitr::kable(instrumentInformation <- do.call('rbind',
  lapply(H, .getInstrumentInformation)), caption = "Instrument information.")
```

## `readIndex` and `readSpectrum`

Defines iRT peptide m/z values and scores.

```{r defineIrtMass}
iRT.mZ <- c(487.2571, 547.2984, 622.8539, 636.8695, 644.8230, 669.8384,
            683.8282, 683.8541, 699.3388, 726.8361, 776.9301)

iRT.score <- c(-24.92, 19.79, 70.52, 87.23, 0, 28.71, 12.39, 33.38, 42.26,
               54.62, 100)

names(iRT.mZ) <- c("LGGNEQVTR", "YILAGVENSK", "GTFIIDPGGVIR", "GTFIIDPAAVIR",
                 "GAGSSEPVTGLDAK", "TPVISGGPYEYR", "VEATFGVDESNAK",
		 "TPVITGAPYEYR", "DGLDAASYYAPVR", "ADVTPADFSEWSK",
		 "LFLQFGAQGSPFLK")
```



```{r defineComputeBestPeptideSpectrumMatch, echo=FALSE}
.computeBestPeptideSpectrumMatch <- function(rawfile,
   pepSeq = "LGGNEQVTR",
   peptideMassTolerance = 0.003,
   fragmentIonMassTolerance = 0.01) {
  
  mass2Hplus <- (protViz::parentIonMass(pepSeq) + 1.008) / 2
  
  S <- rawrr::readIndex(rawfile)
  S <- S[-which(S$MSOrder != "Ms2"), ]
  if (nrow(S) == 0) return (data.frame(scan=NA, peptide=NA,
               rawfile=rawfile, score=NA))

  idx <- lapply(mass2Hplus, function(m){
    which(abs(as.double(S$precursorMass) - m) < peptideMassTolerance)
  })

  # just to be generic correct
  scanNumbers <- lapply(idx, function(x){S$scan[x]})
  
  yIons <- function (b, y) return(y)
  
  bestMatchingMS2Scan <- mclapply(1:length(pepSeq), function(i){
    PL <- rawrr::readSpectrum(rawfile, scan = scanNumbers[[i]])
    
    pp <- lapply(PL, function(x){protViz::psm(pepSeq[i], x, FUN = yIons,
                                              plot = FALSE)})
    
    intensity.max <- max(sapply(pp, function(x) sum(x$intensity)))

## Note: implement some more sophisticated score from the literature
## intensityOfMatches <- sum(PL[[j]]$intensity[abs(pp[[j]]$mZ.Da.error) < 0.1])
## nomalizedIntensityOfMatches 
##   <- median(PL[[j]]$intensity[abs(pp[[j]]$mZ.Da.error) < 0.01]) / intensity.max    
## derive scores for each match
    score <- sapply(1:length(pp),
      function(j){                                                                                                   
        # hard constraint
        nIonHits <- sum(abs(pp[[j]]$mZ.Da.error) < fragmentIonMassTolerance)
        nIonHits
      }) 
    ## determine best scoring spectra
    bestFirstMatch <- which(max(score, na.rm = TRUE) == score)[1]
    
    ## return the winner
    data.frame(scan=scanNumbers[[i]][bestFirstMatch], peptide=pepSeq[i],
               rawfile=rawfile, score=max(score, na.rm = TRUE))
  })
  
  do.call('rbind', bestMatchingMS2Scan)
}
```
As a proof-of-concept, the R function `.computeBestPeptideSpectrumMatch` defines a basic heuristic to determine feasible peptide to spectrum matches (psm) for the iRT peptides.

The snippet below performs the peptide spectrum match number crunching for the `r names(iRT.mZ)[1]` peptide on all raw files utilizing `r detectCores()` cores:

```{r computeBestMatches}
start_time <- Sys.time()
bestMatchingMS2Scan <- mclapply(FUN = .computeBestPeptideSpectrumMatch,
                                input,
                                pepSeq=names(iRT.mZ)[1])
end_time <- Sys.time()
end_time - start_time
```
```{r knitrBestMatches, results='asis'}
S <- do.call('rbind', bestMatchingMS2Scan)
S$Id <- 1:nrow(S)
knitr::kable(S[c("Id", "scan", "score")],
             caption = "Best LGGNEQVTR spectrum matches.")
```

```{r usi, echo=FALSE}
.deriveUniversalSpectrumIdentifier <- function(repoid, rawfile, scan , peptide){
  paste("mzspec", repoid, rawfile, "scan", scan, paste0(peptide, "/2"), sep=":")
}
```


## `plot` spectrum

```{r definePlotBestMatches, echo=FALSE}
.plotBestMatches <- function(x, instrumentModel){
  if(is.na(x$scan)) return (NULL)
  S <- readSpectrum(x$rawfile[1], x$scan)
  
  lapply(1, function(i){
    plot(S[[i]], main=x$peptide[i], sub=instrumentModel)
    Y <- x$peptide[i]
    abline(v=fi<-fragmentIon(Y)[[1]]$y[1:nchar(Y)-1],
           col="#AAAAAA66",lwd=5)
    
    axis(3, fi, paste0('y', 1:length(fi)))
  })}
```


```{r plotBestMatches, fig.show="hold", out.width="50%"}
start_time <- Sys.time()
op <- par(mar = c(6, 6, 6, 1), mfrow=c(1, 1))
rv <- mapply(FUN=.plotBestMatches, bestMatchingMS2Scan,
             instrumentInformation$model)
end_time <- Sys.time()
end_time - start_time
```

## `readChromatogram`

Derive chromatograms for all iRT peptides applying the mass filter option and an ion tolerance of 10ppm:

```{r readChromatogram, error=TRUE}
start_time <- Sys.time()

C <- mclapply(X = input,
              FUN = rawrr::readChromatogram,
              mass = iRT.mZ,
              tol = 10,
              type = "xic",
              filter = "ms")

end_time <- Sys.time()
end_time - start_time
```
## `plot` chromatogram

Visualize the chromatograms and apply a linear regression of the extracted APEX values.

```{r definePlotChromatogramAndFit, echo=FALSE}
.plotChromatogramAndFit <- function(x, i){
    par(mfrow=c(2,1))
    
    plot(x); legend("topright", legend=i, title='Instrument Model', bty = "n", cex=0.75)
    
    rt <- sapply(x, function(x) x$times[which.max(x$intensities)[1]])
    if (length(rt) == length(iRT.score)){
    fit <- lm(rt ~ iRT.score)
    plot(rt ~ iRT.score, ylab = 'Retention time [min]',
         xlab = "iRT score", pch=16, frame.plot = FALSE)
    abline(fit, col = 'grey')
    abline(v = 0, col = "grey", lty = 2)
    legend("topleft",
     legend = paste("Regression line: ", "rt =",
                    format(coef(fit)[1], digits = 4), " + ",
                    format(coef(fit)[2], digits = 2), "score", "\nR2: ",
                    format(summary(fit)$r.squared, digits = 2)),
           bty = "n", cex = 0.75)
    text(iRT.score, rt,  iRT.mZ,pos=1,cex=0.5)
    }
}
```

```{r plotChromatogramAndFit, fig.retina=3,fig.height=9,  fig.show="hold", out.width="50%"}
rv <- mapply(FUN=.plotChromatogramAndFit,
             x=C, i=instrumentInformation$model)
```

\newpage
# Listings of code snippets

The following functions were used in this supplement information report.

## `.getInstrumentInformation`
```{r}
.getInstrumentInformation
```

## `.computeBestPeptideSpectrumMatch`
```{r}
.computeBestPeptideSpectrumMatch
```

## `.plotBestMatches`
```{r}
.plotBestMatches
```

## `.plotChromatogramAndFit`

```{r}
.plotChromatogramAndFit
```

# System information

![rawrrRCMDcheck](https://github.com/fgcz/rawrr/workflows/rawrrRCMDcheck/badge.svg)

## Mono

```{bash monoInfo, message=TRUE, echo=TRUE}
mono --version
```

```{bash    mcs, message=TRUE, echo=TRUE}
mcs --version
```

The following assemblies of the ThermoRawFileReader.4.0.22 version were used:

```
  <ItemGroup>
    <Reference Include="ThermoFisher.CommonCore.Data.dll"/>
    <Reference Include="ThermoFisher.CommonCore.MassPrecisionEstimator.dll"/>
    <Reference Include="ThermoFisher.CommonCore.RawFileReader.dll"/>
 </ItemGroup>
```

## Session information

```{r sessionInfo}
sessionInfo()
```

